bin/main

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+    entry
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+    entry
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    i = 0
+    arr = []
+    while i < size
+      arr << self[i] if yield(self[i])
+      i += 1
+    end
+    arr
+  end
+
+  def my_all?
+    return to_enum unless block_given?
+
+    elem = true
+    i = 0
+    while i < size
+      elem = false unless yield(self[i])
+      i += 1
+    end
+    elem
+  end
+
+  def my_any?
+    return to_enum unless block_given?
+
+    i = 0
+    while i < size
+      return true if yield(self[i])
+
+      i += 1
+    end
+    false
+  end
+
+  def my_none?
+    return to_enum unless block_given?
+
+    elem = true
+
+    i = 0
+    while i < size
+      elem = false if yield(self[i])
+      i += 1
+    end
+    elem
+  end
+
+  def my_count
+    return to_enum unless block_given?
+
+    elem = 0
+
+    i = 0
+    while i < size
+
+      elem += 1 if yield(self[i])
+      i += 1
+    end
+    elem
+  end
+
+  def my_map(arr, prc = nil)
+    elem = []
+
+    i = 0
+    while i < arr.length
+
+      if block_given?
+        yield(arr[i])
+      else
+        elem << prc.call(arr[i])
+      end
+      i += 1
+    end
+    elem
+  end
+
+  def my_inject(arr)
+    elem = arr[0]
+
+    i = 0
+    while i < arr.length
+      elem = yield(elem, arr[i])
+      i += 1
+    end
+    elem
+  end
```

Comment:
* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    operation = proc { |sum, n| sum + n }
    array = [1,2,5,4,7,2]
    p array.my_inject(&operation)
    #returns this error
    #ArgumentError:
    #    wrong number of arguments (given 0, expected 1)
  ```
* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    array = [1,2,5,4,7,2]
    p array.my_inject(:+)
    #returns this error
    #LocalJumpError:
    #    no block given (yield)
  ```